Fix: allow registration of blocks that include assets from within a theme

fabiankaegy opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Allows the registration of blocks using `block.json` to register assets via the `script`, `editorScript`, `viewScript`, `style` or `editorStyle` keys from within a Theme. Previously this only worked from within a plugin / mu-plugin.

Trac ticket: https://core.trac.wordpress.org/ticket/54647 & https://core.trac.wordpress.org/ticket/55513

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

fabiankaegy committed:
allow registration of blocks within themes

fabiankaegy committed:
Merge branch 'trunk' into fix/allow-block-registration-from-within-themes

# Conflicts:
#	src/wp-includes/blocks.php

fabiankaegy committed:
add handeling of stylesheets in theme registered blocks

fabiankaegy committed:
fix use normalized style path

fabiankaegy committed:
handle theme blocks within _wp_multiple_block_styles

fabiankaegy committed:
fix undefined notice by moving style_path_norm into is_theme_block check

gziolo commented on src/wp-includes/blocks.php:
```diff
@@ -108,12 +108,20 @@ function register_block_script_handle( $metadata, $field_name ) {
 	}
 	// Path needs to be normalized to work in Windows env.
 	$wpinc_path_norm  = wp_normalize_path( realpath( ABSPATH . WPINC ) );
+	$theme_path_norm  = wp_normalize_path( get_theme_file_path() );
```
Let’s apply `realpath` for this path as well following other cases. It prevents some mismatches on Windows platform. Unless it’s already the case for `get_theme_file_path`. I couldn’t locate it in the code.

fabiankaegy replied:
I've added the `realpath` wrapper to both instances of the `$theme_path_norm` variable in https://github.com/WordPress/wordpress-develop/pull/2494/commits/679eee0bd0df9553fa72a4da4892e938d488d290

fabiankaegy committed:
fix wrap `get_theme_file_path` with `realpath` to prevent mismatches with windows

gziolo commented:
Can we update the existing unit tests that cover both helper functions to include a check for the plugin and the theme?

https://github.com/WordPress/wordpress-develop/blob/3cd1a741a0c0301f2ddaeee98b86067e94372596/tests/phpunit/tests/blocks/register.php#L241-L253

https://github.com/WordPress/wordpress-develop/blob/3cd1a741a0c0301f2ddaeee98b86067e94372596/tests/phpunit/tests/blocks/register.php#L286-L306

It might not be that simple so I'm happy to help with that.

fabiankaegy committed:
Merge branch 'WordPress:trunk' into fix/allow-block-registration-from-within-themes

fabiankaegy commented:
@gziolo How would you suggest we go about adding the test block to a theme? Should it be a new theme? Or should it be added to an existing one of the test themes? If so which one 🤔 

Also since the check for whether or not a block exists within a theme is reliant on the `get_theme_file_path` path the theme needs to be active in order to be recognized.

gziolo commented:
> @gziolo How would you suggest we go about adding the test block to a theme? Should it be a new theme? Or should it be added to an existing one of the test themes? If so which one 🤔

You can add it to one of the existing test themes. `block-theme` sounds like a great fit:

https://github.com/WordPress/wordpress-develop/tree/trunk/tests/phpunit/data/themedir1/block-theme

Maybe you could add a subfolder `blocks` there and switch to the theme as part of the test. 

> Also since the check for whether or not a block exists within a theme is reliant on the `get_theme_file_path` path the theme needs to be active in order to be recognized.

Example:

https://github.com/WordPress/wordpress-develop/blob/371966a187081662c155ef98ddb7a681dc10f268/tests/phpunit/tests/theme/wpGetGlobalStylesheet.php#L48

It probably requires some changes in the set up:

https://github.com/WordPress/wordpress-develop/blob/371966a187081662c155ef98ddb7a681dc10f268/tests/phpunit/tests/theme/wpGetGlobalStylesheet.php#L10

fabiankaegy committed:
add test case for registering blocks within a theme

fabiankaegy commented:
@gziolo I've added two initial test cases following the example of the already existing tests. However, I'm not quite sure what I would need to change here in order to be able to change the theme within the test: https://github.com/WordPress/wordpress-develop/blob/371966a187081662c155ef98ddb7a681dc10f268/tests/phpunit/tests/theme/wpGetGlobalStylesheet.php#L10

Would love to get some more insight into what is required to get that working :)

fabiankaegy committed:
fix remove invalid comment

fabiankaegy committed:
fix add rtl stylesheet to theme block test case

fabiankaegy commented:
(force pushed to re-run the GitHub actions as per [this Slack conversation](https://wordpress.slack.com/archives/C02RQBWTW/p1649073538260479?thread_ts=1649073113.733219&cid=C02RQBWTW))

fabiankaegy commented:
@gziolo Seems that even without any changes to the setup the tests have passed

gziolo commented on tests/phpunit/tests/blocks/register.php:
```diff
@@ -252,6 +252,22 @@ public function test_success_register_block_script_handle() {
+			'name'   => 'block-theme/example-block',
+			'script' => 'file:./index.js',
+		);
+		$result   = register_block_script_handle( $metadata, 'script' );
```
So this one tres to register the `script` while the theme reigsters the `editorScript` but both use the same filename, right?

fabiankaegy replied:
Updated this to reference `viewScript` in the test

gziolo commented on tests/phpunit/tests/blocks/register.php:
```diff
@@ -305,6 +321,28 @@ public function test_success_register_block_style_handle() {
+			'name'  => 'block-theme/example-block',
+			'style' => 'file:./style.css',
+		);
+		$result   = register_block_style_handle( $metadata, 'style' );
```
So this one tres to register the `style` but the theme reigsters the `style` in `block.json` and both use the same filename, right?

fabiankaegy replied:
updated this to reference `editorStyle` in the test

gziolo commented:
> @gziolo Seems that even without any changes to the setup the tests have passed

I think, we should make it explicit that the theme and tests register different scripts and styles so we could ensure that their paths are different. Such check is missing for the script. Otherwise, everything looks good.

fabiankaegy committed:
update test specs to reference other assets than block.json

fabiankaegy committed:
fix allow arrow indentation in test cases

gziolo commented on tests/phpunit/tests/blocks/register.php:
```diff
@@ -252,6 +252,22 @@ public function test_success_register_block_script_handle() {
+		);
+		$result   = register_block_script_handle( $metadata, 'viewScript' );
+
+		$this->assertSame( 'block-theme-example-block-view-script', $result );
```
Can we also check the `src` set for the file? You should be able to get it with `wp_scripts()->registered[$name]` or  `wp_scripts()->query( $name )`.

gziolo replied:
It looks like the URL generated for the block is invalid. See the URL produced in the test:

<img width="1838" alt="Screenshot 2022-04-06 at 16 13 50" src="https://user-images.githubusercontent.com/699132/161995165-94ccf614-7876-45ee-9b33-9d65af115655.png">

fabiankaegy replied:
Okay that is odd. It is working fine when I actually test it manually with different themes that have blocks with assets. I wonder whether there is something with how the active theme is set in these tests? 🤔 

Do you have any insight there? / any idea about it?

gziolo reviewed (approved):
Awesome work. I think we have all we need. Let's ensure that all CI checks pass.

gziolo commented on tests/phpunit/tests/blocks/register.php:
```diff
@@ -305,6 +321,28 @@ public function test_success_register_block_style_handle() {
+		$this->assertSame(
+			wp_normalize_path( realpath( DIR_TESTDATA . '/themedir1/block-theme/blocks/example-block/editor-style.css' ) ),
+			wp_normalize_path( wp_styles()->get_data( 'block-theme-example-block-editor-style', 'path' ) )
+		);
```
The same issue with the style URL when I extend this test:

<img width="1852" alt="Screenshot 2022-04-06 at 16 17 50" src="https://user-images.githubusercontent.com/699132/161996031-1b02768d-16f5-410f-ae45-5671713f863c.png">

gziolo committed:
Update register.php

fabiankaegy committed:
fix use correct file path for metadata file in theme test cases

fabiankaegy committed:
fix use correct theme file path in theme test case

fabiankaegy committed:
add debugging code to debug CI test results

fabiankaegy committed:
fix update block ckecks to use normalized script path

fabiankaegy committed:
Revert "fix update block ckecks to use normalized script path"

This reverts commit d32c73d1bbbc38321b77527baa25860a3da565e9.

fabiankaegy committed:
fix update is_theme_block check to use normalized path

fabiankaegy committed:
add more debugging values

fabiankaegy committed:
remove debugging code from branch

fabiankaegy committed:
try remove realpath check from theme_path_norm variable

fabiankaegy committed:
update is_theme_block check for stylesheets to use normalized path

fabiankaegy committed:
fix linting issue

fabiankaegy committed:
fix exclude broken test cases

These tests are brobek because the test runner setup does not stub out the theme uri functions and therefore the stylesheet uri does not match the expectations.

These tests were added as part of this PR but because we cannot get them to work in CI we are removing them for now

fabiankaegy committed:
fix remove broken test case

fabiankaegy commented:
As part of this PR @gziolo and myself tried to add more robust tests for the assets that get registered via `block.json` files in the theme. 

We've tried to actually test the src url of the registered script instead of only the handle:
```php
$expected_script_handle = 'block-theme-example-block-view-script';
$this->assertSame( 'http://example.org/wp-content/themes/block-theme/blocks/example-block/view.js', wp_scripts()->query( $expected_script_handle )->src );
```

However the test setup does not seem to have stubs in place for the `get_theme_file_uri` function calls which resulted in the tests passing successfully when running the test suite locally but failing when the tests were run in CI. 

After some considerable time debugging we have decided to forgo the addition of these additional tests for now and they have been removed from the PR. 

However it would be nice to further stub out the various theme functions so that testing themes gets a little easier and that we can more confidently test theme functionality.

gziolo reviewed (approved):
I tested it with my local copy of the WordPress development version with the patch applied with `npm run grunt patch:https://github.com/WordPress/wordpress-develop/pull/2494`.

I can confirm that all core blocks load as before. I tested it also with the custom plugin created with `@wordprtess/create-block` using the same template as in the tutorial in Block Editor Handbook. Everything works as expected.

It looks that either the unit tests environment needs some tweaks or the way we switch the test block theme needs to get adjusted. The idea of adding additional assertions was to cover the edge cases we recently fixed for core blocks and plugins in https://core.trac.wordpress.org/ticket/55311.

gziolo commented:
Committed with https://core.trac.wordpress.org/changeset/53091.

jack-h2ml commented:
My apologies for commenting well after the fact, please do let me know if this is not an appropriate course of action. 

The implementation above does not work within the Parent / Child theme paradigm. The changes introduced within [wp-includes/blocks.php](https://core.trac.wordpress.org/changeset/53091/trunk/src/wp-includes/blocks.php) do not cover the use case whereby a block may be registered within a parent theme, to be used by subsequent child themes.  

I have raised an issue covering this in more detail over on the Gutenberg repo WordPress/gutenberg/#47470
